packages/core/lib/queues/AbstractQueueService.ts

```diff
@@ -216,10 +201,51 @@ export abstract class AbstractQueueService<
         messageId,
       )
     }
+
+    if (!this.logMessages && !this.messageMetricsManager) {
+      return
+    }
+
+    const processedMessageMetadata = this.resolveProcessedMessageMetadata(
+      message,
+      processingResult,
+      messageId,
+    )
     if (this.logMessages) {
-      // @ts-ignore
-      const resolvedMessageId: string | undefined = message?.[this.messageIdField] ?? messageId
-      this.logProcessedMessage(message, processingResult, resolvedMessageId)
+      this.logger.debug(
+        processedMessageMetadata,
+        `Finished processing message ${processedMessageMetadata.messageId}`,
+      )
+    }
+    if (this.messageMetricsManager) {
+      this.messageMetricsManager.registerProcessedMessage(processedMessageMetadata)
+    }
+  }
+
+  protected resolveProcessedMessageMetadata(
+    message: MessagePayloadSchemas | null,
+    processingResult: MessageProcessingResult,
+    messageId?: string,
+  ): ProcessedMessageMetadata {
+    // @ts-ignore
+    const resolvedMessageId: string | undefined = message?.[this.messageIdField] ?? messageId
+
+    const messageTimestamp = message ? this.tryToExtractTimestamp(message) : undefined
+    const messageProcessingMilliseconds = messageTimestamp
+      ? Date.now() - messageTimestamp.getTime()
```

Comment:
I would set constant to `Date.now` as early as possible, to make the duration calculation more accurate

Reply:
Adjusted